pinns/src/pinns.c

```diff
@@ -268,20 +294,62 @@ static bool is_host_ns(const char* const optarg) {
   return optarg && !strcmp (optarg, HOSTNS);
 }
 
+// Verify the /$PATH/$NSns directory exists and is a directory, returning the
+// expanded path name in bind_root and returning the length of bind_root (or -1
+// on error)
+static int create_bind_root(char *bind_root, size_t size, const char *pin_path, const char *ns_name) {
+  int len = snprintf(bind_root, size, "%s/%sns", pin_path, ns_name);
+  if (directory_exists_or_create(bind_root) < 0) {
+    pwarnf("%s exists and is not a directory", bind_root);
+    return -1;
+  }
+  return len;
+}
+
+/* Mount namespaces can only be bound into unsharable mount namespaces (to
+ * avoid infinite loops), so force the pin_path to be a bind-mount to
+ * itself and then marked as unsharable. */
+static int setup_unbindable_bindpath(const char *pin_path, const char *ns_name) {
+  char bind_root[PATH_MAX];
+
+  // first, verify the /$PATH/$NSns directory exists
+  if (create_bind_root(bind_root, PATH_MAX - 1, pin_path, ns_name) < 0) {
+    return -1;
+  }
```

Comment:
can these functions go beneath their last caller?

Reply:
ah I see setup_unbindable_bindpath is, but create_bind_root could be moved to the bottom of the file

Reply:
Will do.